runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/ExecutionStateTracker.java

```diff
@@ -298,7 +304,17 @@ public long getNextLullReportMs() {
     return nextLullReportMs;
   }
 
-  protected void takeSample(long millisSinceLastSample) {
+  void takeSample(long millisSinceLastSample) {
+    if (SAMPLING_UPDATER.compareAndSet(this, 0, 1)) {
```

Comment:
Do we need sampling/SAMPLING_UPDATER because you wanted to be able to have multiple threads activating/deactivating states within a single tracker at the same time?

Note there is only one ExecutionStateSampler thread and each tracker is associated with a single bundle at a time. To emulate this behavior we would want to have one sampler that is shared by N trackers each with their own states where N is the number of concurrent threads you want to use.

Reply:
There's a race now between `removeTracker` removing the tracker from the concurrent set (and them sampling it) and the sampling thread enumerating it and sampling it.  Previously the synchonrized block would prevent that race.  This is essentially pushing the lock down to the individual tracker instances to prevent it being sampled concurrently.

Reply:
Are you saying that there is another thread that does the sampling that isn't the ExecutionStateSampler thread?

Reply:
Correct, it's sampled in `removeTracker` as well https://github.com/apache/beam/blob/534e99eae37e34760baa7beb432a18dc4cd148a5/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/ExecutionStateSampler.java#L166

Reply:
Thanks, I missed that.